MSC1708: .well-known support for server name resolution (#1708)
matrix-org/matrix-doc: proposals/1708-well-known-for-federation.md

  endpoints or to failover to another endpoint if an endpoint fails.

The first two points remain unchanged: if the server name is an IP literal, or
contains a port, then requests will be made directly as before.

If the hostname is neither an IP literal, nor does it have an explicit port,
then the requesting server should continue to make an SRV lookup as before, and
use the result if one is found.

If *no* result is found, the requesting server should make a `GET` request to
`https://\<server_name>/.well-known/matrix/server`, with normal X.509
certificate validation. If the request fails in any way, then we fall back as
before to using using port 8448 on the hostname.

Rationale: Falling back to port 8448 (rather than aborting the request) is
necessary to maintain compatibility with existing deployments, which may not
present valid certificates on port 443, or may return 4xx or 5xx errors.

If the GET request succeeds, it should result in a JSON response, with contents
structured as shown:

```json
{
   "server": "<server>[:<port>]"
}
```

The `server` property has the same format as a [server
name](https://matrix.org/docs/spec/appendices.html#server-name): a hostname

> followed by an optional port.

Reviewer: This sounds like `server` is going to end up being the exact same as the request in the first place (ie: searching for `example.org` results in a `"server": "example.org"`). What's the expected value here? How do we prevent routing loops? What sort of processing should be done on the resulting name?

Author: Yup, I'll rephrase this to make it obvious it's a different thing.

If the response cannot be parsed as JSON, or lacks a valid `server` property,
the request is considered to have failed, and no fallback to port 8448 takes
place.

Otherwise, the requesting server performs an `AAAA/A` lookup on the hostname,
and connects to the resultant address and the specifed port. The port defaults
to 8448, if unspecified.

### Caching

Servers should not look up the `.well-known` file for every request, as this
would impose an unacceptable overhead on both sides. Instead, the results of
the `.well-known` request should be cached according to the HTTP response
headers, as per [RFC7234](https://tools.ietf.org/html/rfc7234). If the response
does not include an explicit expiry time, the requesting server should use a
sensible default: 24 hours is suggested.

Because there is no way to request a revalidation, it is also recommended that
requesting servers cap the expiry time. 48 hours is suggested.

Similarly, a failure to retrieve the `.well-known` file should be cached for

> a reasonable period. 24 hours is suggested again.

Reviewer: we should lower the suggestion to be 12 hours or less imo. There's going to be the vast majority of people who don't have the .well-known file, but we also shouldn't cause too much pain for people trying to troubleshoot or fix their federation. In a scenario where a server they have no control over caches the admin's typo, federation is broken for a day. This makes it difficult to figure out if your change is actually going to help or not.

Author: Yeah, I'd had the same thoughts tbh. 12h sgtm.

Author: actually: it should probably be even shorter. What we don't want is a single 500 error or networking blip to take out federation to a server for ages.

(I wonder if we should consider trying to re-validate a cached .well-known well before it expires, so we can have a few goes at fetching it before it actually expires.)

Reviewer: Could probably lower this all the way down to an hour. Negotiating some sort of retry schedule might work, although I suspect that if someone screwed up their config then they'd screw that up too by accident.

Reviewer 2: I think we should be a bit cautious about this. If someone (forcefully) deprovisions a matrix server then having a large number of servers requesting `.well-known` forever more at a high rate is problematic (a thousand servers requesting every hour is quite a few requests)

Author: @erikjohnston does this concern go away a bit if we cache 404s for a bit longer (say 24 hours)? I was thinking this short validity period should only apply to 500s and connection failures.

Reviewer: could probably cache a 410 for a whole lot longer than a 404 if people really want to distance themselves from Matrix forever.

Reviewer 3: @erikjohnston is it possible to use this .well-known file to decommission a homeserver? (It would be great to solve https://github.com/matrix-org/synapse/issues/3765 in this proposal if that's possible)

> # MSC1708: .well-known support for server name resolution

Reviewer 2: (As asked, moved from issue comment:)

Thanks for writing this up! Having read through it my main concerns with this approach are:

1. This is yet another look up mechanism. As more servers are implemented will it be too easy for them to end up with subtle inconsistencies, e.g. doing `.well-known` look ups first?
2. We're reinventing the wheel, in terms of basically wanting to use `.well-known` for the same use as what use SRV records were designed for. Instead of being able to reuse what DNS provides we need to have conversations for each individual feature (e.g. caching, load balancing, priorities, etc).

One alternative to this approach is for the `.well-known` to record the fact that the domain wants to *delegate* the handling of their Matrix server to a separate entity (rather than have use it as a look up). For example, if `foo.com` delegated to `foo.bar.com` in its `.well-known` then other matrix servers would first check the `.well-known` for the delegation and then do a lookup via SRV for `foo.bar.com`.

This has the advantages that:

1. It still allows servers to delegate their domain to another entity, without having to hand over their certificates.
2. Doesn't reinvent some of the nicer features of SRV records. (Including allowing the hosting entity to set up secondary backup matrix servers a la MX records.)
3. Potentially allows the `.well-known` look up to be cached more aggressively, if it is assumed that changing delegations happens less frequently than re-pointing DNS. This may or may not be a bit spurious.

The drawback is that this obviously doesn't help folks who a) can't add SRV records and b) can't just use 8448 on their domain.

Author: > 1. This is yet another look up mechanism. As more servers are implemented will it be too easy for them to end up with subtle inconsistencies, e.g. doing .well-known look ups first?

Possibly, although I'm not sure we avoid that with your suggestion either. (It's a shame that DNS lookups are hard to test with sytest; which, incidentally, is a good argument against SRV in my book.)

> 1. It still allows servers to delegate their domain to another entity, without having to hand over their certificates.

(to be clear: the problem is with the keys, not the certs; the certs are of course public.)

> 2. Doesn't reinvent some of the nicer features of SRV records. (Including allowing the hosting entity to set up secondary backup matrix servers a la MX records.)

Yeah, though as I've suggested above, I'm not convinced that backup servers are that much use as a feature. 

> 3. Potentially allows the .well-known look up to be cached more aggressively, if it is assumed that changing delegations happens less frequently than re-pointing DNS. This may or may not be a bit spurious.

I'm not sure how this really helps. I think we can cache `.well-known` results for long enough as it is, and can use the `Cache-Control` header to do so. The problem comes when there is no `.well-known` result - and we'd have that problem whether or we did an SRV lookup after the `.well-known` lookup or not.

Reviewer 2: > Yeah, though as I've suggested above, I'm not convinced that backup servers are that much use as a feature.

I don't really see why not, backup SMTP servers are common enough (even matrix.org has one). While matrix has a few more reliability mechanisms, its not clear to me that people wouldn't want to set up a backup store and forward server for matrix (and I don't think it would be particularly hard to implement)

Author: I just feel like we might be over-designing this when nobody has yet written a store-and-forward server.

Reviewer 4: > I just feel like we might be over-designing this when nobody has yet written a store-and-forward server.

Nobody has yet written a spec (even whatever is documented) compliant server, and the spec did not allow anyone else to write one except NV as they had the tribal knowledge. So I feel this is a premature assertion and that the cause & effect is misunderstood.
By implementing a homeserver myself (and soon two), it's not that there is a lack of interest in a multi-server setup, there definitely is. It's just that it's not possible to implement a server at this point.

Maybe this should be revisited after S2S r0 is out and enough time was given for an implementation to possibly arise?

Reviewer 4: > We see MSC1711 as essential for S2S r0, and we can't do that without solving the domain-delegation problem.

That's fine, but this is MSC1708 which seems to have a hard-dependency on MSC1711 (but was written before?). Do you mean MSC1708 is essential for S2S r0?

Reviewer 5: As pointed out above, Priority and Weight have lots of usecases. It would be a shame if we went from a standard lookup method that supported those back to something where we need to figure out basically everything from scratch (as Erik pointed out) where we get less features (which would be easy to implement here too).

Reviewer 2: This entire discussion is somewhat my point in terms of wanting to avoid even having to have the discussion about what we want `.well-known` to support :)

---

> > 3. Potentially allows the .well-known look up to be cached more aggressively, if it is assumed that changing delegations happens less frequently than re-pointing DNS. This may or may not be a bit spurious.
>
> I'm not sure how this really helps. I think we can cache .well-known results for long enough as it is, and can use the Cache-Control header to do so. The problem comes when there is no .well-known result - and we'd have that problem whether or we did an SRV lookup after the .well-known lookup or not.

Hmm, true. Ignore that then.

---

I'm somewhat struggling to figure out *why* I feel like simply stating a delegation is a better approach here. It somehow feels cleaner to be more explicit on what is trying to be achieved and then using a single approach to resolve IPs.

There is also a clearer interaction with TLS certificates and Host headers. If `.well-known` is basically another way of resolving IPs then it feels a bit awful that `.well-known` connections use different Host/SNI/TLS certs than when looking up via SRV, assuming MSC1711. (From an implementation perspective this is a bit icky since it'd be a bit harder to abstract the lookup away entirely).
If the `.well-known` had a delegation in it, then it makes more sense that the Host/SNI/etc are pointing to the server that is being delegated to. 

One nice feature that does come out of having explicit delegations is that if server `foo.com` delegated to `bar.com`, then whenever a server tried to send an event to to both they could simply send the event *once* to `bar.com`. This would be valuable for deployments doing vhosting.

Reviewer 2: To summarise, the flow would be:
1. Check for a delegation for in the `/.well-known` for the given server name
2. Send requests for the original server to the delegated server in the same way as  today, (i.e. look up the SRV records etc).

The advantages that I see are:
1. We don't have to worry about what features of SRV we want to re-implement in `.well-known` (that includes time spent wondering whether we need any of those features or not)
2. Doesn't have the same oddness of using different Host/SNI/TLS certs depending on whether the lookup found a `.well-known` or not.
3. Differentiates between the idea of an (active) HS and a "virtual" server that simply has users on but offloads all the actual work to a HS. I.e. you always talk to an actual HS, and always use it for the Host/SNI/TLS etc, but still allows using a non-HS domain in user IDs etc.

These are all quite hand wavy, but it does feel to me like its a cleaner solution somehow. 

I also quite like the idea that if we end up in a world where user IDs are public keys─and so don't encode the routing information─we would only have to do the `/.well-known` look up when resolving aliases like `@user:example.com` or `#room:example.com` (on the assumption that when users joined a room the routing information would point to the delegated server rather than the original server). This also means that things don't go weird for existing conversations if the `/.well-known` gets broken, though on the flip side does make it more complicated to change the delegation (but that is always going to be a bit messy).

Reviewer 5: The problem I see with that is that it is a breaking change in how to look up hosts. Going with a verification policy served via .well-known means that only the TLS validation changes. (Proper writeup with lookup flow is still pending, university stuff got in the way)

Author: > 1. We don't have to worry about what features of SRV we want to re-implement in .well-known (that includes time spent wondering whether we need any of those features or not)

Agreed.

> 2. Doesn't have the same oddness of using different Host/SNI/TLS certs depending on whether the lookup found a .well-known or not.

I'm not sure I consider this an oddness as it stands in the current MSC, but in any case it doesn't feel like you avoid it with your proposal.

> 3. Differentiates between the idea of an (active) HS and a "virtual" server that simply has users on but offloads all the actual work to a HS. I.e. you always talk to an actual HS, and always use it for the Host/SNI/TLS etc, but still allows using a non-HS domain in user IDs etc.

I kinda see what you mean here, but am not entirely convinced it's such an improvement that it buys us a great deal.

Having said all that, I think my only real argument against doing an SRV after a .well-known lookup is the added network traffic/latency in the (presumably unusual) case that a .well-known is present. As others have pointed out, the .well-known is going to dominate here anyway.

Currently, mapping from a server name to a hostname for federation is done via
`SRV` records. This presents two principal difficulties:

>  * SRV records are not widely used, and administrators may be unfamiliar with

Reviewer 5: When I got into selfhosting, `.well-known` was more confusing to me than SRV records. I had no problem getting my synapses up and running, but I still haven't gotten around to putting proper .well-known records for my WebDAV host, which I could have done instantly if I had understood how.

   them, and there may be other practical difficulties in their deployment such

>    as poor support from hosting providers. [^1]

Author: @jcgruenhage is this intended as an answer to @maxidor ? it seems somewhat tangential (and I don't agree with it, but please start another thread if you want to discuss.)

@maxidor: I'm not sure I follow you. `.well-known` is just a file on a webserver - it's hard for a hosting provider to mess that up. SRV records on the other hand are something that DNS providers can and do fail to correctly support.

>    as poor support from hosting providers. [^1]

Reviewer 5: It was meant as an addition to what @maxidor said. In the case of Cloudflare this is obviously a bit more complicated, I didn't take that into consideration.

>    as poor support from hosting providers. [^1]

Reviewer 4: @richvdh I think the bottom line question is: Is this proposal made to solve the shortcomings of a single DNS provider for a single domain, or an actual widespread issue? Because across all the domains I have ever managed, my personal experience with SRV is that it never failed to work or be supported for all the registars/DNS providers I've used (OVH, Gandi, Online, Proximus, GoDaddy, AWS, etc.).

>    as poor support from hosting providers. [^1]

Author: @maxidor the problem is not just with providers: it is also about whether people trying to deploy synapses understand SRV records. People like you and jcg may know all about them, but I'm willing to believe that more people can deploy a file on their webserver than set up an SRV record.

Ultimately though, this point is somewhat moot. The driver here is the delegation problem (below).

>    as poor support from hosting providers. [^1]

Reviewer 5: Just as I said earlier, when I started to get into this back a few years ago, SRV records were easier to grasp than .well-known stuff.

 * It is likely that we will soon require valid X.509 certificates on the
   federation endpoint. It will then be necessary for the homeserver to present
   a certificate which is valid for the server name. This presents difficulties
   for hosted server offerings: BigCorp may be reluctant to hand over the
   keys for `bigcorp.com` to the administrators of the `bigcorp.com` matrix

>    homeserver.

Reviewer 4: The section doesn't explain how a possible change in X.509 certificate requirement has any relevance with SRV vs well-known. The relevance is also not clear.

Reviewer 5: The idea here was that we can't trust DNS without DNSSEC, so if a SRV of the server `bigcorp.com` points to `mitm-agent.com`, we still require the cert to be valid for bigcorp.com. When using `.well-known`, the record is served via HTTPS, so this is already authenticated and we can trust a certificate only valid for the hostname specified in the `.well-known` record. IMO the solution to this is to require DNSSEC when pointing to a different host (instead of just a different port)

Reviewer 4: Something still doesn't add up to me... For basic certificates, nearly all CA's rely on a DNS entry to issue a certificate. Commercial CAs either use a CNAME, SRV or a TXT record (like AWS). Let's encrypt uses an HTTP(S) connection to the resolved hostname.

Knowing that, this would be the "chain of trust" for
- SRV: DNS lookup for domain -> SRV record -> DNS lookup for SRV hostname -> Connection to HS
- well-known: DNS lookup for domain -> HTTPS connection to domain -> Certificate validation -> Connection to HS

In both case, the weak link is DNS and in both case, you can MitM as you can edit the SRV record or automatically get a certificate for a domain you can hijack. One is not more secure than the other.

As this section is about self-signed vs CA-issues certs (correct me if I'm wrong), this is why I don't understand the relevance of having a self-signed or a CA-issued certificate. In both case, MitM would happen the same way and the type of certificate makes no difference.

Reviewer 5: The difference here is that for getting a valid TLS certificate, you need to MitM the DNS of both the user and the CA, instead of just the user. An attacker on a public network might be able to modify the DNS requests sent in that network, but not the network of the CA. 

This section is also not about self-signed vs CA-issued certs. The decision here is whether we require that the TLS cert of the homeserver is valid for server.name or host.name.that.server.name.points.to. If the authenticity of the pointer isn't guaranteed, we need the TLS cert to be valid for server.name. `.well-known` is authenticated, because it has to be served through HTTPS with a valid TLS cert. SRV records are only authenticated when DNSSEC is used.

Reviewer 4: I think splitting up the section into the various points would help:
> It is likely that we will soon require valid X.509 certificates on the federation endpoint.

This talks about a potential change, in the future, where the `valid` conditions are not spelled out. On the other hand, the spec already [has something about this](https://matrix.org/docs/spec/server_server/unstable.html#authentication).
It's hard to make any kind of judgement if we don't know the difference between now (which is also unclear) and later. But let's assume the next sentence is the only condition.

> It will then be necessary for the homeserver to present a certificate which is valid for the server name.

This is already partially the case in practice: synapse checks that the certificate is issued for the Matrix domain. What synapse doesn't check is that the fingerprint is matching what is presented at `GET /_matrix/key/v2/server/{keyId}`.
These are the only two elements that can be checked with the current federation and what someone implementing a Homeserver would assume (synapse does, construct does, gridepo does, Dendrite possibly does).

So at this point, it's only a matter of clarification and I still don't see a relevant point that sits in favor of adding .well-know complexity to federation discovery.

> This presents difficulties for hosted server offerings: BigCorp may be reluctant to hand over the keys for `bigcorp.com` to the administrators of the `bigcorp.com` matrix homeserver.

They don't need to or have to. The federation endpoint can be reverse proxy. This is a fundamental infrastructure setup which affects any kind of service dealing with this (think email, xmpp, etc) - hardly isolated to Matrix. It might be a good idea to see how other protocols solved it?

If we look at existing implementations that are live in a production public environment (synapse only), the private key of the certificate is not required. The minimum required is the fingerprint of the certificate, which can be automatically fetched in recent versions unless I'm mistaken.

So all in all, I'm still not understanding why needing a valid certificate, which in practice is/should be required anyway, is an argument to add .well-known support and/or removing SRV records.

Author: MSC1711 suggests that we should do the normal CA checks on TLS certificates instead of the business with tls_fingerprints. (Sorry not to have linked that MSC previously, and sorry if you missed it).

That introduces the problem that the current way of delegating responsibility for running the homeserver for a domain - the SRV record - will require the domain admins to hand over the private keys for the domain. That problem is solved by adding `.well-known` support.

And yes, there alternative solutions to that problem, such as requiring those domain admins to set up a reverse proxy, or DNSSEC. The idea of this proposal is to try to flesh out the way that a `.well-known` would work so we can try to evaluate it.

Author: > This is already partially the case in practice: synapse checks that the certificate is issued for the Matrix domain.

If only this were true :/.

Reviewer 4: > That introduces the problem that the current way of delegating responsibility for running the homeserver for a domain - the SRV record - will require the domain admins to hand over the private keys for the domain. That problem is solved by adding .well-known support.

Could you give a concrete example with concrete domain names and what DNS & HTTP traffic would look like? I'm having a hard time understanding what you mean by "private key" as I explained above why TLS private key is not a hard requirement.

Author: Google want to try out Matrix and want to out-source running a matrix server (to Modular, for sake of an example): 

Option 1 (based on SRV records):
 * `_matrix._tcp.google.com` points to `google.modular.im`
 * Requesting servers expect `google.modular.im` to present an SSL cert for `google.com`
 * so Google has to either (1) give modular the private keys for their existing cert , or (2) participate in a certificate signing process to get a new certificate for `google.com`, using keys which Modular control. Either way Modular end up with an authenticated certificate for serving `google.com`. (They can then use this certificate to MITM all of Google's other traffic, or something.)

Option 2 (based on .well-known):
 * `https://google.com/.well-known/matrix/server` points to `google.modular.im`
 * Requesting servers now expect `google.modular.im` to present an SSL cert for `google.modular.im`
 * There is no problem with certificates. Google has not had to hand over their crown jewels to test out a chat service.

>  TLS private key is not a hard requirement.

Unless I'm missing something spectacular, it is a hard requirement.

Reviewer 5: @maxidor the point here is though that for intercepting matrix traffic, you just need to be able to fake DNS request. If the certificate for the original name is required, then the verification will fail.

Reviewer 4: @jcgruenhage My original comment is not on a section related to DNS trust or verification, but on the algo used to verify that the TLS certificate is valid. This is not related to DNS validation but to TLS hostname validation.

Author: > I see the problem and also why this is a problem: that's not how you're supposed to validate the TLS certificate of SRV record 

Says who? And if you do it the other way, how does that not drive a massive DNS-spoofing-shaped security hole through the whole thing? (So yes, it is absolutely related to DNS validation, or rather the absence thereof)

Reviewer 4: > Says who?

What I am really saying here and in some other comments about the other sections is that what we are discussing here is not uncommon and has been solved several times in the past. Validating a certificate hostname for a SRV record is done daily by other technologies.
Validating that DNS records have not been spoofed is done daily using technologies that exists for a long time now, and awareness of the insecurity of DNS there.

So to repeat what hasn't been addressed or replied to: Check how other technologies in the same position are doing you don't reinvent the wheel. Other techs solve this by using DNSSEC and a check of the target hostname of the SRV record.

Author: citation required.

Author: [we aren't going to require dnssec. period.]

Reviewer 4: > citation required.

Seems like I forgot to reply to this. Just check how Emails are done in regards of SMTP STARTLS when dealing with MX records per example, which follow a similar delegation system as SRVs.
Let me show you with two obvious examples: Emails to `gmail.com` and `hotmail.com` (I kept the relevant output of commands)

### Gmail
```
$ dig -t MX +short gmail.com
5 gmail-smtp-in.l.google.com.  # We gonna use this one
30 alt3.gmail-smtp-in.l.google.com.
40 alt4.gmail-smtp-in.l.google.com.
10 alt1.gmail-smtp-in.l.google.com.
20 alt2.gmail-smtp-in.l.google.com.

$ openssl s_client -starttls smtp -connect gmail-smtp-in.l.google.com:25
CONNECTED(00000003)
depth=2 OU = GlobalSign Root CA - R2, O = GlobalSign, CN = GlobalSign
verify return:1
depth=1 C = US, O = Google Trust Services, CN = Google Internet Authority G3
verify return:1
depth=0 C = US, ST = California, L = Mountain View, O = Google LLC, CN = mx.google.com
verify return:1
---
Certificate chain
 0 s:/C=US/ST=California/L=Mountain View/O=Google LLC/CN=mx.google.com
   i:/C=US/O=Google Trust Services/CN=Google Internet Authority G3
 1 s:/C=US/O=Google Trust Services/CN=Google Internet Authority G3
   i:/OU=GlobalSign Root CA - R2/O=GlobalSign/CN=GlobalSign
---
Server certificate
-----BEGIN CERTIFICATE-----
[...]
-----END CERTIFICATE-----
subject=/C=US/ST=California/L=Mountain View/O=Google LLC/CN=mx.google.com
issuer=/C=US/O=Google Trust Services/CN=Google Internet Authority G3
[...]
Verification: OK
[...]
    Verify return code: 0 (ok)
---
250 SMTPUTF8
^C
```
The certificate has `mx.google.com` has main subject, and contains the MX hostname in the alternative names, but **not** `gmail.com`. Feels free to check for yourself.

### hotmail.com
```
# dig -t MX +short hotmail.com
2 hotmail-com.olc.protection.outlook.com.

# openssl s_client -starttls smtp -connect hotmail-com.olc.protection.outlook.com:25
CONNECTED(00000003)
depth=2 C = BE, O = GlobalSign nv-sa, OU = Root CA, CN = GlobalSign Root CA
verify return:1
depth=1 C = BE, O = GlobalSign nv-sa, CN = GlobalSign Organization Validation CA - SHA256 - G3
verify return:1
depth=0 C = US, ST = Washington, L = Redmond, O = Microsoft Corporation, CN = mail.protection.outlook.com
verify return:1
---
Certificate chain
 0 s:/C=US/ST=Washington/L=Redmond/O=Microsoft Corporation/CN=mail.protection.outlook.com
   i:/C=BE/O=GlobalSign nv-sa/CN=GlobalSign Organization Validation CA - SHA256 - G3
 1 s:/C=BE/O=GlobalSign nv-sa/CN=GlobalSign Organization Validation CA - SHA256 - G3
   i:/C=BE/O=GlobalSign nv-sa/OU=Root CA/CN=GlobalSign Root CA
---
Server certificate
[...]
subject=/C=US/ST=Washington/L=Redmond/O=Microsoft Corporation/CN=mail.protection.outlook.com
issuer=/C=BE/O=GlobalSign nv-sa/CN=GlobalSign Organization Validation CA - SHA256 - G3
---
[...]
Verification: OK
---
[...]
    Verify return code: 0 (ok)
    Extended master secret: yes
---
250 SMTPUTF8
^C
```
The certificate has `mail.protection.outlook.com` has main subject, and contains `*.olc.protection.outlook.com` as alternative name that matches the MX hostname in the alternative names. It does **not** contain `hotmail.com`. Again, check for yourself.

This works the same way with XMPP and other services that rely on SRV (or MX which is a primal SRV). Nobody ever uses the original hostname with SRV/MX when validating a cert. You use the hostname specificed in the SRV/MX target because this is not a CNAME. This is an explicit indication that "my service is delivered by X", and not "My canonical name is X".

It would be good if the core team would avoid trying to reinvent the wheel. If your concern is DNS spoofing, then activate DNSSEC. This is done by the DNS admin and is automatically handled by intermediary DNS cache up to the client requester.

If you think DNS is just insecure and don't want to use it for things like this, you're better off dropping DNS altogether everywhere and create something that would replace it, like everyone in P2P networks are trying to do.

Reviewer 6: @jcgruenhage In your opinion is this proposal predicated on the idea that matrix server operators won't run DNSSEC, and as a system administrator yourself can you tell us if DNSSEC is an unreasonable burden administrators?

While there's something to be said for engineering a system that is idiot-proof by default, is it really necessary, or should we think so poorly of the matrix ecosystem as to hold their youthful hands by this .well-known rather than letting them use DNSSEC (something which there are other reasons to use and everyone should be anyway)

Reviewer 5: I don't think DNSSEC is much of a burden, and my experience has been very painless. I have heard of very severe problems though, which can render your whole zone and everything below that unusable for *days*.

About the foolproof part: You can still run a secure matrix server perfectly fine without DNSSEC, if you give a certificate  with the server name in a SAN entry to it.

Reviewer 5: I don't think DNSSEC is much of a burden, and my experience has been very painless. I have heard of very severe problems though, which can render your whole zone and everything below that unusable for *days*.

About the foolproof part: You can still run a secure matrix server perfectly fine without DNSSEC, if you give a certificate  with the server name in a SAN entry to it.

Reviewer 6: @jcgruenhage Would you say that HTTP/3 (QUIC) would mitigate most of the latency round-trip concerns with this proposal? Do you believe New Vector is capable of pushing QUIC in synapse and dendrite congruent to this proposal inevitably landing?

Reviewer 5: HTTP/3 wold definitely help (if I can count it's 5 round trips (A/AAAA for .well-known -> QUIC for .well-known -> SRV -> A/AAAA -> QUIC) instead of 11), but it's not released yet, it's not going to be released that soon, and even once the specification is fixed getting everyone to upgrade is going to take forever. The problems with people on ancient twisted versions exist already, getting the majority of people onto a twisted version that supports QUIC will probably take too long. Dendrite is a different story, getting QUIC support into that shouldn't be a problem at all. Apart from that, if we make the jump to QUIC but don't do .well-known, we get down to 3 roundtrips (SRV -> A/AAAA -> QUIC), which would be really nice. 

Less relevant sidenote: While DNS is definitely less of a problem than TCP/TLS/HTTP roundtrip wise, if QUIC happens then DNS is the slowest part of this chain. Getting rid of it would be kinda nice :)

Reviewer 6: One other thing I'd like some clarification on: Does this proposal constitute a breaking change? It appears so from what I've read. In other words, once this is introduced it *must* be used. If a server does not conduct the `.well-known` query as a step in the overall process it very well may never be able to contact a server and the federation will have additional fragmentation. This is somewhat bothersome. I always like to consider what the forward-compatibility lifespan is for code written (i.e imagine if Construct or all the other ecosystem servers were now complete and their principal developers have moved on). Now in one fell swoop everything breaks and has to be revisited to work again. Is there any mitigation for this which the spec can contribute? If not, is it possible for an implementation to only check .well-known as a *last resort* after otherwise being unsuccessful in contacting a server?

>    homeserver.

Author: Sadly no. For a big corporate IT department, setting up DNSSEC is going to be even scarier than handing over the keys to their master domain.

>    homeserver.

Reviewer 5: After checking adoption among companies that I'd consider to have a "big corporate IT department", I guess you're right. I couldn't find a single massive corp that uses DNSSEC

>    homeserver.

Reviewer 4: @jcgruenhage which domains did you check exactly?

>    homeserver.

Reviewer 5: telekom.com, microsoft.com, google.com, siemens.com, ibm.com, redhat.com, amazon.com, facebook.com

>    homeserver.

Reviewer 5: Sure, this is not a complete list, but instead a small list of random companies that came to my mind.

Here we propose to solve these problems by augmenting the current `SRV` record
with a `.well-known` lookup.

## Proposal

For reference, the current [specification for resolving server
names](https://matrix.org/docs/spec/server_server/unstable.html#resolving-server-names)
is as follows:

* If the hostname is an IP literal, then that IP address should be used,
  together with the given port number, or 8448 if no port is given.

* Otherwise, if the port is present, then an IP address is discovered by
  looking up an AAAA or A record for the hostname, and the specified port is
  used.

* If the hostname is not an IP literal and no port is given, the server is
  discovered by first looking up a `_matrix._tcp` SRV record for the
  hostname, which may give a hostname (to be looked up using AAAA or A queries)
  and port.  If the SRV record does not exist, then the server is discovered by
  looking up an AAAA or A record on the hostname and taking the default
  fallback port number of 8448.

  Homeservers may use SRV records to load balance requests between multiple TLS

>   endpoints or to failover to another endpoint if an endpoint fails.

Author: I'm just quoting from the existing spec...

>   endpoints or to failover to another endpoint if an endpoint fails.

Reviewer 5: Okay

The first two points remain unchanged: if the server name is an IP literal, or
contains a port, then requests will be made directly as before.

If the hostname is neither an IP literal, nor does it have an explicit port,
then the requesting server should continue to make an SRV lookup as before, and
use the result if one is found.

If *no* SRV result is found, the requesting server should make a `GET` request
to `https://\<server_name>/.well-known/matrix/server`, with normal X.509
certificate validation. If the request fails in any way, then we fall back as
before to using using port 8448 on the hostname.

Rationale: Falling back to port 8448 (rather than aborting the request) is
necessary to maintain compatibility with existing deployments, which may not
present valid certificates on port 443, or may return 4xx or 5xx errors.

If the GET request succeeds, it should result in a JSON response, with contents
structured as shown:

```json
{

>    "server": "<server>[:<port>]"

Reviewer 4: Is the lack of namespacing in the key name intentional?

Author: Not particularly. I couldn't see any particular reason for namespacing it though.

Author: >  it'd be good to have something in there to indicate that it was actually meant as a matrix thing

well, that was the thinking behind having `/matrix` in the URL. isn't that good enough for this?

Reviewer 2: Not if people are forwarding all of `.well-known/` to something that then just returns a JSON blob.

Author: > Not if people are forwarding all of .well-known/ to something that then just returns a JSON blob.

... don't do that, then?

Reviewer 2: Well, ok. Though I'd have thought being a bit paranoid and namespacing it would be a quick way of avoid any pain down the line.

Reviewer: +1 to namespacing. At the very least to remain consistent with the client-server .well-known.

Author: have added `m.`, though I'm still unclear what it buys us.

}

> ```

Reviewer 4: +1

> ```

Author: I thought about this. Given that (afaik) there aren't currently any server implementations which could plausibly be deployed in a multi-homed environment such that failover and load-balancing would be useful features, it seemed a bit pointless, so I decided to KISS rather than get bogged down trying to imagine what a suitable configuration might look like for a hypothetical future implementation.

We can always add another key in the future if we want to support load-balancing. (Servers which don't support load-balancing will fall back to the current `server` key, and nothing of value will be lost)

> ```

Reviewer 5: That isn't really true, Synapse runs on boxes with multiple (maybe unreliable) uplinks, where failover would be really useful. I'd like to run in such a setup, but https://github.com/matrix-org/synapse/issues/4104 currently blocks that. A workaround here would be to have a server on the web with a fixed IP that does NAT with failover/loadbalancing over VPN connections, but that increases the latency and decreases the bandwidth, increasing the cost.

> ```

Reviewer 5: [Dendrite](https://github.com/matrix-org/dendrite/blob/262fc25aca9c81d0014754c05ea1ab237e614964/vendor/src/github.com/matrix-org/gomatrixserverlib/client.go#L114) and Construct (couldn't find line here) also don't handle SRV records correctly though

> ```

Reviewer 6: > Construct

[The line is right here](https://github.com/jevolk/charybdis/blob/af93d55927aa7d2614e71e9b66b80e77942f5ab9/modules/s_dns.cc#L126)

The weight and the priority from multiple records are simply not factored, but they're all in that vector; multiple A records are similarly incomplete. 

As a one man show, having been forced to implement a custom DNS client to support asynchronous SRV queries (to support SRV _at all_ for cross-platform unices) I have to budget my tasks appropriately. Even though making use of multiple records is only a few dozen more lines of code, I saw (and currently see) a very tiny number of deployments (we're talking about matrix.org and maybe two or three others) which return multiple DNS records.

When a matrix server makes, or is close to making, any real use of multi-homing I will find that to be the appropriate time to complete those few lines. As this is not the case: the current incomplete state was easier to create, and easier to maintain and debug for the time being.

Feel free to contribute to Construct to add support for features like this: idle and perform [#test:zemos.net](https://matrix.to/#/#test:zemos.net)

> ```

Reviewer 5: Especially in a decentralized mesh world, this isn't all that niche. SRV records would work perfectly for this, just as they would for a properly set up HA matrix server. Making SRV records work would certainly not be designing around this niche feature.

The `server` property should be a hostname or IP address, followed by an
optional port.

If the response cannot be parsed as JSON, or lacks a valid `server` property,
the request is considered to have failed, and no fallback to port 8448 takes
place.

Otherwise, the requesting server performs an `AAAA/A` lookup on the hostname
(if necessary), and connects to the resultant address and the specifed
port. The port defaults to 8448, if unspecified.

(The formal grammar for the `server` property is identical to that of a [server
name](https://matrix.org/docs/spec/appendices.html#server-name).)

### Caching

> Servers should not look up the `.well-known` file for every request, as this

Reviewer: one thing to consider is that servers may very well ignore this. I'm assuming that servers have the right to rate limit / prevent DoS flooding on the .well-known route, therefore potentially causing the remote end (who is disrespecting the rules) to be a sad pile of sad.

Author: well, sure.

would impose an unacceptable overhead on both sides. Instead, the results of
the `.well-known` request should be cached according to the HTTP response
headers, as per [RFC7234](https://tools.ietf.org/html/rfc7234). If the response
does not include an explicit expiry time, the requesting server should use a
sensible default: 24 hours is suggested.

Because there is no way to request a revalidation, it is also recommended that
requesting servers cap the expiry time. 48 hours is suggested.

Similarly, a failure to retrieve the `.well-known` file should be cached for
a reasonable period. 24 hours is suggested again.

### The future of SRV records

It's worth noting that this proposal is very clear in that we will maintain
support for SRV records for the immediate future; there are no current plans to
deprecate them.

However, clearly a `.well-known` file can provide much of the functionality of
an SRV record, and having to support both may be undesirable. Accordingly, we

> may consider sunsetting SRV record support at some point in the future.

Reviewer 6: Contradictory statements like these aren't helpful to implementations as their authors might think. This a technical specification. It's requiring that SRV be supported. It's obvious that this well-known gimmick provides redundant functionality, and it's apparent from this paragraph that the authors are disappointed in themselves for their preceding mistakes costing implementations which now have to support and maintain twice as much to get a single feature value.

But that sprinkle of doubt in that last sentence: that it will all be wasted effort, maybe soon, maybe never -- really salts the wound.

Please.

Reviewer 5: A `.well-known` file also a speciality that matrix homeservers would need to implement, while SRV records are a ***standard*** that lots of applications use. Their use is widespread enough that registrars and DNS hosters really should fix it if broken.

Reviewer 5: A `.well-known` file also a speciality that matrix homeservers would need to implement, while SRV records are a ***standard*** that lots of applications use. Their use is widespread enough that registrars and DNS hosters really should fix it if broken.

Author: > Their use is widespread enough that registrars and DNS hosters really should fix it if broken.

That may be the case, but the fact that Cloudflare have been promising to do so for nearly a year and haven't done so isn't promising IMHO. To me it reflects the fact that SRVs are (rightly or wrongly) seen as exotic functionality, whereas a `.well-known` is just a web server serving a file, so there's less to go wrong.

Reviewer 5: Reading the CF forum thread you linked further down I'd say they think their behaviour isn't wrong, since they most stuff SRV records are used for are things they don't proxy. Another workaround they've proposed in the thread is to have two different domains in cloudflare and then have the SRV record on one of the domains point to a host that is proxied on the other domain, which will then not be flattened. Still annoying, but it works and looks like it is intentional, so I wouldn't necessarily call the current behaviour broken.

Author: We tried that and failed to get it to work. I don't know the details now tbh, and it may have been a confusion. (@ara4n any ideas here?)

Either way, it still makes SRV records a bit of a foot-gun.

Reviewer 7: To my knowledge we never tried having two different CF domains to avoid flattening (e.g. SRV record for matrix.org pointing to a record on matrix2.org to avoid flattening).  Instead, (ab)using the `matrix.org.cdn.cloudflare.net` record as a target in order to avoid flattening worked.

### Outstanding questions

Should we follow 30x redirects for the .well-known file? On the one hand, there
is no obvious usecase and they add complexity (for example: how do they
interact with caches?). On the other hand, we'll presumably be using an HTTP
client library to handle some of the caching stuff, and they might be useful
for something?

## Problems

It will take a while for `.well-known` to be supported across the ecosystem;
until it is, it will be difficult to deploy homeservers which rely on it for
their routing: if Alice is using a current homeserver implementation, and Bob
deploys a new implementation which relies on `.well-known` for routing, then
Alice will be unable to send messages to Bob. (This is the same problem we have with
[SNI](https://github.com/matrix-org/synapse/issues/1491#issuecomment-415153428).)

The main defence against this seems to be to release support for `.well-known`

> as soom as possible, to maximise uptake in the ecosystem. It is likely that, as

Reviewer 8: s/soom/soon/

we approach Matrix 1.0, there will be sufficient other new features (such as
new Room versions) that upgrading will be necessary anyway.

## Security considerations

The `.well-known` file potentially broadens the attack surface for an attacker
wishing to intercept federation traffic to a particular server.

## Conclusion

This proposal adds a new mechanism, alongside the existing `SRV` record lookup
for finding the server responsible for a particular matrix server_name, which
will allow greater flexibility in deploying homeservers.

> [^1] For example, Cloudflare automatically "flattens" SRV record responses.

Reviewer 5: Could you elaborate what this means and what problems that imposes?

Author: Sure. The idea of Cloudflare is obviously that it sits in front of your origin, and the true origin IP address is obscured (and you really don't want to leak your origin IP to a potential DDoSer).

So: you tell Cloudflare to proxy `matrix.org`, which means that A/AAAA lookups on `matrix.org` resolve to Cloudflare's servers (and also means that you have to let Cloudflare run the whole of your DNS :( ). 

Then you tell Cloudflare that you want `_matrix._tcp.matrix.org` to point to `matrix.org`. But when you then look up `_matrix._tcp.matrix.org`, you get some other hostname which resolves to ... the origin IP!

See https://community.cloudflare.com/t/error-maybe-by-design-in-creating-srv-records/39207 for some other people complaining about this.

Reviewer 5: Okay, wow. That sounds like a horrible bug in their flattening. Can't you disable flattening for SRV records, like it's possible for CNAME records?

Author: No you can't disable it. It's not obvious if they consider this a bug or not.

one).

This MSC proposes to solve this problem by augmenting the current `SRV` record
with a `.well-known` lookup.

## Proposal

For reference, the current [specification for resolving server
names](https://matrix.org/docs/spec/server_server/unstable.html#resolving-server-names)
is as follows:

1. If the hostname is an IP literal, then that IP address should be used,
   together with the given port number, or 8448 if no port is given.

2. Otherwise, if the port is present, then an IP address is discovered by
   looking up an AAAA or A record for the hostname, and the specified port is
   used.

3. If the hostname is not an IP literal and no port is given, the server is
   discovered by first looking up a `_matrix._tcp` SRV record for the
   hostname, which may give a hostname (to be looked up using AAAA or A queries)
   and port.

4. Finally, the server is discovered by looking up an AAAA or A record on the
   hostname, and taking the default fallback port number of 8448.

We insert the following between Steps 3 and 4.

If the SRV record does not exist, the requesting server should make a `GET`
request to `https://<server_name>/.well-known/matrix/server`, with normal X.509

> certificate validation, and following 30x redirects. If the request does not

Reviewer 8: should be careful about blindly following 30x redirects, as servers could get caught in a redirect loop.  In practice, servers will probably want to limit the number of redirects that they'll follow, though I don't know if it's worth mentioning in the spec.  Limiting the number of redirects seems to be common practice in several HTTP request libraries already.

return a 200, continue to step 4, otherwise:

The response must have a `Content-Type` of `application/json`, and must be
valid JSON which follows the structure documented below. Otherwise, the
request is aborted.

If the response is valid, the `m.server` property is parsed as
`<delegated_server_name>[:<delegated_port>]`, and processed as follows:

* If `<delegated_server_name>` is an IP literal, then that IP address should be
  used, together with `<delegated_port>`, or 8448 if no port is given. The
  server should present a valid TLS certificate for `<delegated_server_name>`.

> * Otherwise, if the port is present, then an IP address is discovered by

Reviewer 8: I found the "Otherwise, if" a bit confusing as I read the "if the port is present" as trying to be the negation of the previous if statement rather than an "else if".
```suggestion
* If `<delegated_server_name>` is not an IP literal, and the `<delegated_port>` is present, then an IP address is discovered by
```

(The existing "Resolving server names" section also has the same issue, so maybe that can be fixed there too.)

  looking up an AAAA or A record for `<delegated_server_name>`, and the
  specified port is used. The server should present a valid TLS certificate
  for `<delegated_server_name>`.

  (In other words, the federation connection is made to
  `https://<delegated_server_name>:<delegated_port>`).

* If the hostname is not an IP literal and no port is given, a second SRV
  record is looked up; this time for `_matrix._tcp.<delegated_server_name>`,
  which may give yet another hostname (to be looked up using A/AAAA queries)
  and port. The server must present a TLS cert for the
  `<delegated_server_name>` from the .well-known.

* If no SRV record is found, the server is discovered by looking up an AAAA
  or A record on `<delegated_server_name>`, and taking the default fallback
  port number of 8448.

  (In other words, the federation connection is made to
  `https://<delegated_server_name>:8448`).

### Structure of the `.well-known` response

The contents of the `.well-known` response should be structured as shown:

```json
{
   "m.server": "<server>[:<port>]"
}
```

> The `m.server` property should be a hostname or IP address, followed by an

Reviewer 8: "should be" seems a bit weak here
```suggestion
The `m.server` property is a hostname or IP address, followed by an
```

optional port.

> If the response cannot be parsed as JSON, or lacks a valid `server` property,

Reviewer 8: ```suggestion
If the response cannot be parsed as JSON, or lacks a valid `m.server` property,
```

the request is considered to have failed, and no fallback to port 8448 takes
place.

> (The formal grammar for the `server` property is identical to that of a [server

Reviewer 8: ```suggestion
(The formal grammar for the `m.server` property is identical to that of a [server
```

Though this parenthetical feels like it should be moved two paragraphs up, which also talks about the format of the `m.server` field.
